### ISSUE TITLE
[Bug]: Removing egress/ingress blocks from aws_security_group does not remove the rules

### ISSUE THREAD

**Issue by grimm26** [hashicorp/terraform-provider-aws#31004]
### Terraform Core Version

1.4.5

### AWS Provider Version

4.64.0

### Affected Resource(s)

aws_security_group

### Expected Behavior

rules that were created by the egress and ingress blocks would be destroyed when the blocks were removed.

### Actual Behavior

no change in state or reality.

### Terraform Configuration Files

```hcl
provider "aws" {}

resource "aws_security_group" "test" {
  name_prefix = "funtimes"
  description = "testing stuff"
  vpc_id        = "vpc-98jhafd8gh" #put a valid VPC id

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
  egress {
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    cidr_blocks = ["10.0.0.0/32"]
  }

  ingress {
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    cidr_blocks = ["10.0.0.0/32"]
  }
}
```

### Steps to Reproduce

1. apply the included config to create a security group with rules created by ingress and egress blocks.
2. comment out the ingress and egress blocks.
3. apply again.

### Would you like to implement a fix?

No

**Comment by justinretzolk** [comment]
Hey @grimm26 👋 Thank you for taking the time to raise this! In this case, `ingress` and `egress` are both processed in [attribute-as-blocks mode](https://developer.hashicorp.com/terraform/language/attr-as-blocks), which means they behave in a bit of a unique manner. The following quote from the linked article relates to what you're seeing here:

> However, in some cases existing provider features were relying on the conflation of these two concepts in the language of Terraform v0.11 and earlier, using nested block syntax in most cases but using argument syntax to represent explicitly the idea of removing all existing objects of that type, since the absence of any blocks was interpreted as "ignore any existing objects".

What this means for your scenario is that by removing all of the `ingress` and `egress` blocks, you're hitting the "ignore and existing objects" path. If you would like to remove all of the `ingress` and `egress` rules, you may do so by explicitly setting `ingress = []` and `egress = []`, rather than removing those arguments from the configuration entirely.

**Comment by grimm26** [comment]
@justinretzolk Aha! That helps. I was trying to migrate to using aws_vpc_security_group_ingress_rule and aws_vpc_security_group_egress_rule resources instead. There is now hope! :)

**Comment by grimm26** [comment]
I spoke too soon. So, setting `egress = []` in the `aws_security_group` resource deleted the rule set previously by a block. I also add aws_vpc_security_group_egress_rule blocks. The problem is that the subsequent `apply` will delete the new egress resources since they are tracked inside the `aws_security_group` even when they are created by outside resources.

I could add a lifecycle rule to ignore changes in egress, but only after it deleted the orginal egress in the first apply. Any other ideas?

**Comment by justinretzolk** [comment]
Hey @grimm26 👋 I've not yet had a chance to test this, but if my understanding of the `aws_security_group` resource is correct, after you've moved over to the `aws_vpc_security_group_egress_rule`, you should be able to remove the `egress` argument from the `aws_security_group` resource entirely, and it should switch back to the behavior that you initially reported -- ignoring the egress rules entirely. Is that something you've already tested?

**Comment by grimm26** [comment]
@justinretzolk yes that is the first thing I did. It fails because that does not remove the egress rule and blocks me from creating the same rule with the aws_security_group_egress_rule.

**Comment by justinretzolk** [comment]
Hey @grimm26 -- perhaps I'm misunderstanding, or failed to address a key point in my previous comments. It sounds like the steps you took were to remove the `egress` blocks from the `aws_security_group`, which meant that resource was no longer handling the management of the egress rule(s) (but notably did not remove them).  After that, you attempted to use the `aws_security_group_egress_rule` to take over managing the egress rules, which failed given that they currently exist.

If I understood that correctly, the thing that I'd failed to mention was the step of [importing](https://registry.terraform.io/providers/-/aws/latest/docs/resources/vpc_security_group_egress_rule#import) the `aws_security_group_egress_rule` before attempting a new plan/apply.

**Comment by grimm26** [comment]
@justinretzolk yeah my strategy ended up being setting a lifecycle rule for `ignore_changes` on `egress` and `ingress` for the `aws_security_group` and importing the egress rule into an `aws_security_group_egress_rule` resource. I was hoping to get by without state manipulation, but that's it goes sometimes.
